### ISSUE TITLE
Memory leaks

### ISSUE THREAD

**Issue by wigior** [embedthis/goahead#142]
### Environment
GoAhead version: 3.4.7, built from source
OS: Ubuntu 14.04 x86-64
Hardware: PC
Customization: None

Build using:
```
make
```

Run with test/web as document root.

### Problem Description
When repeating a GET of /auth/digest/digest.html with digest authentication, the memory usage keeps increasing:
```
<<< Request
GET /auth/digest/digest.html HTTP/1.1
User-Agent: curl/7.35.0
Host: localhost
Accept: */*
goahead: 2: Access Denied. User not logged in.

>>> Response
HTTP/1.1 401 Unauthorized
Server: GoAhead-http
Date: Wed Jun 17 10:56:09 2015
WWW-Authenticate: Digest realm="example.com", domain="127.0.1.1", qop="auth", nonce="MGI4YzVlYmU2ODU4OTU3NmYzOWU1NGE0ZDEzMzNhMmY6ZXhhbXBsZS5jb206NTU4MTUyNDk6MA==", opaque="5ccc069c403ebaf9f0171e9517f40e41", algorithm="MD5", stale="FALSE"
Content-Length: 151
Connection: keep-alive
Content-Type: text/html

<<< Request
GET /auth/digest/digest.html HTTP/1.1
Authorization: Digest username="unknown", realm="example.com", nonce="MGI4YzVlYmU2ODU4OTU3NmYzOWU1NGE0ZDEzMzNhMmY6ZXhhbXBsZS5jb206NTU4MTUyNDk6MA==", uri="/auth/digest/digest.html", cnonce="M2Q4ZjZhMDY3MmEyZTcxNTAwMDAzZmVhMDAwYjgwOTU=", nc=00000001, qop=auth, response="442cce46a54ed4cca69759c44ed577fc", opaque="5ccc069c403ebaf9f0171e9517f40e41", algorithm="MD5"
User-Agent: curl/7.35.0
Host: localhost
Accept: */*
goahead: 2: Access denied: user is unknown
goahead: 2: Access Denied. User not logged in.

>>> Response
HTTP/1.1 401 Unauthorized
Server: GoAhead-http
Date: Wed Jun 17 10:56:09 2015
WWW-Authenticate: Digest realm="example.com", domain="127.0.1.1", qop="auth", nonce="MGI4YzVlYmU2ODU4OTU3NmYzOWU1NGE0ZDEzMzNhMmY6ZXhhbXBsZS5jb206NTU4MTUyNDk6MQ==", opaque="5ccc069c403ebaf9f0171e9517f40e41", algorithm="MD5", stale="FALSE"
Content-Length: 151
Connection: keep-alive
Content-Type: text/html
```

### Analysis
The parseDigestDetails function allocates memory for a copy of the authentication details. This is never freed.
```
auth.c: 781: key = sclone(wp->authDetails);
```

The websError function allocates memory for a log message. This is never freed.
```
http.c: 1769: msg = sfmtv(fmt, args);
```

The termWebs function contains conditionally compiled code to free the digest authentication details. This should use ME_GOAHEAD_DIGEST instead of ME_DIGEST.
```
http.c: 485: #if ME_DIGEST
```

**Comment by embedthis** [comment]
Both proposed fixes confirmed. Proposed patch:

```
diff --git a/src/auth.c b/src/auth.c
index cd21c03..4a2402c 100644
--- a/src/auth.c
+++ b/src/auth.c
@@ -774,11 +774,11 @@ static void digestLogin(Webs *wp)
 static bool parseDigestDetails(Webs *wp)
 {
     WebsTime    when;
-    char        *decoded, *value, *tok, *key, *dp, *sp, *secret, *realm;
+    char        *decoded, *value, *tok, *key, *keyBuf, *dp, *sp, *secret, *realm;
     int         seenComma;

     assert(wp);
-    key = sclone(wp->authDetails);
+    key = keyBuf = sclone(wp->authDetails);

     while (*key) {
         while (*key && isspace((uchar) *key)) {
@@ -906,6 +906,8 @@ static bool parseDigestDetails(Webs *wp)
             }
         }
     }
+    wfree(keyBuf);
+
     if (wp->username == 0 || wp->realm == 0 || wp->nonce == 0 || wp->route == 0 || wp->password == 0) {
         return 0;
     }
diff --git a/src/http.c b/src/http.c
index 8987ed9..8474064 100644
--- a/src/http.c
+++ b/src/http.c
@@ -482,7 +482,7 @@ static void termWebs(Webs *wp, int reuse)
 #if ME_GOAHEAD_CGI
     wfree(wp->cgiStdin);
 #endif
-#if ME_DIGEST
+#if ME_GOAHEAD_DIGEST
     wfree(wp->cnonce);
     wfree(wp->digestUri);
     wfree(wp->opaque);
```

**Comment by wigior** [comment]
There are three issues that cause a memory leak. The proposed patch fixes only two of them.

**Comment by embedthis** [comment]
Thanks. Quick patch for the missing item:
```
diff --git a/src/http.c b/src/http.c
index 8474064..1d9ed4d 100644
--- a/src/http.c
+++ b/src/http.c
@@ -1765,11 +1765,12 @@ PUBLIC void websError(Webs *wp, int code, char *fmt, ...)
     wfree(wp->url);
     wp->url = encoded;
     if (fmt) {
-        va_start(args, fmt);
-        msg = sfmtv(fmt, args);
-        va_end(args);
         if (!(code & WEBS_NOLOG)) {
+            va_start(args, fmt);
+            msg = sfmtv(fmt, args);
+            va_end(args);
             trace(2, "%s", msg);
+            wfree(msg);
         }
         buf = sfmt("\
 <html>\r\n\
```

**Comment by embedthis** [comment]
Fixed in 3.4.9
